rfcs/0008-storage-handler.md

```diff
@@ -0,0 +1,338 @@
+
+## Storage Slot Assignment
+
+Each storage variable in a `storage` block is assigned an index corresponding to its location in the list of variables:
+
+```rust
+struct MyInnerStruct {
+    c: u64,
+    d: b256,
+}
+
+struct MyStruct {
+    a: u64,
+    b: MyInnerStruct,
+}
+
+storage {
+    x: u64 = 0, // Assigned index `0`
+    y: b256 = ZERO_B256, // Assigned index `1`
+    str: str[99], // Assigned index '2'
+    s: MyStruct = MyStruct { a: 0, b: MyInnerStruct { c: 0, d: ZERO_B256 } }, // Assigned index `3`
+}
+```
+
+The "index" assigned to each variable is used to generated a key as follows.
+
+### "Small" Types
+
+Small types are types that fit in a single storage slot. We exclude structs with multiple fields from this definition even they do fit in one slot; structs with at least one field are considered "large types" are handled in the [next section](#large-types).
+
+Storage variables that have small types are assigned a single key that is equal to `sha256("storage_<idx>")` where `<idx>` is the index assigned to the storage variable. In the example above, the key chosen for `x` is:
+
+```rust
+sha256("storage_0") = 0xf383b0ce51358be57daa3b725fe44acdb2d880604e367199080b4379c41bb6ed
+```
+
+and the key chosen for `y` is:
+
+```rust
+sha256("storage_1") = 0xde9090cb50e71c2588c773487d1da7066d0c719849a7e58dc8b6397a25c567c0
+```
+
+### "Large" Types
+
+Large types span multiple storage slots and are laid out sequentially in storage starting at key `sha256("storage_<idx>")` where `<idx>` is the index assigned to each storage variable. The number of storage slots required can be computed as `(__size_of::<T>() + 31) >> 5` (ceiling of the size of the data type divided by `32`). Structs are the only exception to this rule because the fields of a struct are each stored in their own storage slot regardless of their size. This makes accessing or writing a particular struct field much more efficient.
+
+> **Note**
+> Struct variables are laid out in the same way today.
+
+For example, the storage variable `str` above is a string containing `99` characters and its size is `99` bytes which span 2 storage slots. It is also assigned index `2`. Therefore, `str` spans slots with keys `sha256("storage_2")` and `sha256("storage_2") + 1`.
+
+Variable `s` on the other hand, is a nested struct and is assigned index `3`. Therefore, each of its fields (and subfields) is assigned a separate key starting with `sha256("storage_3")`:
+
+- `s.a` is assigned `sha256("storage_3")`.
+- `s.b.c` is assigned `sha256("storage_3") + 1`.
+- `s.b.d` is assigned `sha256("storage_3") + 2`.
+
+### Empty Types
+
+There are two possible empty types: structs with no fields and enums with no variants. Empty structs are useful for implementing dynamic storage types such as `StorageMap` and `StorageVec`. Storage variables that have empty types should not consume any storage slots.
+
+### Behavior of `StorageKey`
```

Comment:
Since we're defining `StorageKey` behavior, it seems storing values at different hashes is inefficient. The current behavior creates the need for an additional 32 bytes for each storage variable, which increases all contract sizes. The alternative would be to hash the slot at runtime, adding additional gas to already expensive storage reads/writes.

For comparison, Solidity increments starting at slot zero and increments by one for each variable. Dynamic arrays store their length in the slot and the array starts at `keccak256(slotIndex) + arrayIndex` and hash maps are stored at `keccak256(abi.encodePacked(key, slotIndex))`.

This minimizes bytecode size since the slots start at zero and only use hash digests for dynamic values.

Reply:
The main reason why packing was not originally considered is because it leads to read-modify-write patterns. Say we have
```
storage {
    x: u64 = 0,
    y: u64 = 0,
    z: u64 = 0,
    w: u64 = 0,
}
```
and we decide to store all the variables in a single storage slots (say at key 0x00..0) since they actually fit there. If I would like to update `x`, then I need to first **read** from `0x00.00`, do some shifting/masking to update only the bytes that `x` occupy, and then **write** back the result. Isn't more expensive than just writing `x` directly when it's in its own slot?

Reply:
One thing Im not sure about: Are we charged by the storage slot or by the byte when reading/writing ?
Normally I would assume the former, but I seem to recall some discussion about allowing for reading multiple slots in one read for example.

Reply:
Packing is now the default but only for structs.